File: backend-javascript/02-backend-javascript-prototype/src/modules/profession/profession.repository.mock.js
import { ITEMS_MOCK_DATA } from '../../mocks/profession/profession.mock-data.js';
import createItem from './profession.model.js';

class MockRepository {
  constructor() {
    this.items = JSON.parse(JSON.stringify(ITEMS_MOCK_DATA));
  }

  async getItems({ offset = 0, limit = 10 } = {}) {
    const total = this.items.length;
    const totalPages = Math.ceil(total / limit);
    const data = this.items.slice(offset, offset + limit);

    const metadata = {
      totals: {
        currentPageTotals: {
          count: data.length,
          offset: offset,
          limit: limit,
        },
        globalTotals: {
          count: total,
          totalPages: totalPages,
        },
      },
    };

    return { metadata, data };
  }

  async getItemsCount() {
    return { count: this.items.length };
  }

  async getItemById(id) {
    return this.items.find((item) => item.id === id) || null;
  }

  async createItem(data) {
    const newItem = createItem({ id: this.items.length + 1, ...data });
    this.items.push(newItem);

    return newItem;
  }

  async updateItem(id, data) {
    const index = this.items.findIndex((item) => item.id === id);
    if (index === -1) { return null; }
    this.items[index] = { ...this.items[index], ...data };

    return this.items[index];
  }

  async deleteItem(id) {
    const index = this.items.findIndex((item) => item.id === id);
    if (index === -1) { return null; }

    return this.items.splice(index, 1)[0];
  }

  async existsByName(name) {
    return this.items.some(
      item => item.name.toLowerCase() === name.toLowerCase(),
    );
  }
}

export default MockRepository;

File: backend-java/03-backend-java/src/main/java/com/ganatan/mocks/ProfessionRepositoryMock.java
package com.ganatan.mocks;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import java.util.Optional;

import com.ganatan.modules.profession.Profession;

public class ProfessionRepositoryMock {

    private final List<Profession> items = new ArrayList<>();
    private int nextId = 1;

    public ProfessionRepositoryMock() {
        items.add(new Profession(nextId++, "Actor Backend Mock"));
        items.add(new Profession(nextId++, "Actress Backend Mock"));
        items.add(new Profession(nextId++, "Animator Backend Mock"));
        items.add(new Profession(nextId++, "Art Director Backend Mock"));
        items.add(new Profession(nextId++, "Associate Producer Backend Mock"));
        items.add(new Profession(nextId++, "Cinematographer Backend Mock"));
        items.add(new Profession(nextId++, "Composer Backend Mock"));
        items.add(new Profession(nextId++, "Costume Designer Backend Mock"));
        items.add(new Profession(nextId++, "Editor Backend Mock"));
        items.add(new Profession(nextId++, "Executive Producer Backend Mock"));
    }

    public Map<String, Object> getItems(int offset, int limit) {
        int totalItems = items.size();
        int totalPages = (int) Math.ceil((double) totalItems / limit);
        int currentPage = (offset / limit) + 1;
        int fromIndex = Math.min(offset, totalItems);
        int toIndex = Math.min(offset + limit, totalItems);
        List<Profession> paginatedItems = items.subList(fromIndex, toIndex);

        Map<String, Object> pagination = new HashMap<>();
        pagination.put("currentPage", currentPage);
        pagination.put("perPage", limit);
        pagination.put("totalItems", totalItems);
        pagination.put("totalPages", totalPages);

        Map<String, Object> metadata = new HashMap<>();
        metadata.put("pagination", pagination);

        Map<String, Object> response = new HashMap<>();
        response.put("metadata", metadata);
        response.put("data", paginatedItems);

        return response;
    }

    public Optional<Profession> getItemById(int id) {
        return items.stream().filter(item -> item.getId() == id).findFirst();
    }

    public Profession createItem(Profession profession) {
        Profession newItem = new Profession(nextId++, profession.getName());
        items.add(newItem);
        return newItem;
    }

    public Optional<Profession> updateItem(int id, Profession updatedData) {
        Optional<Profession> existing = getItemById(id);
        existing.ifPresent(item -> item.setName(updatedData.getName()));
        return existing;
    }

    public boolean deleteItem(int id) {
        return items.removeIf(item -> item.getId() == id);
    }

    public boolean existsByName(String name) {
        return items.stream().anyMatch(item -> item.getName().equalsIgnoreCase(name));
    }
}
